### PR TITLE
MINOR: Speed up JoinGroupRequestTest, ListGroupsRequestTest, SyncGroupRequestTest by decreased  rebalance deplay from 3s to 1s

### PR DESCRIPTION
**Pull request by m1a2st**
Due to this test is too slow, it can't test in 30 seconds. Thus I add a config to speed up this test, let it can completed in 30s

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala [hunk 1]
```diff
@@ -36,13 +36,14 @@ import scala.concurrent.duration.Duration
...
-@Timeout(120)
+@Timeout(30)
 @ExtendWith(value = Array(classOf[ClusterTestExtensions]))
 class JoinGroupRequestTest(cluster: ClusterInstance) extends GroupCoordinatorBaseRequestTest(cluster) {
   @ClusterTest(types = Array(Type.KRAFT), serverProperties = Array(
     new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "true"),
     new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
-    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1"),
+    new ClusterConfigProperty(key = "group.initial.rebalance.delay.ms", value = "0"),
```
Comment (lianetm) [inline core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala]: Interesting finding, makes sense to me for the test. I expect it would it be helpful to add it also in the other test in this file, right?

Comment (m1a2st) [reply core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala]: You are right, but we test these tests without add  `group.initial.rebalance.delay.ms` and find a problem that Kraft mode is obviously slower than Zookeeper mode, thus we guess that root cause is the metadata synchronization speed.
螢幕快照 2024-08-06 21-14-26 (1)

**Comment by dajac** [comment]
@m1a2st Thanks for the patch. I understand the motivation but I wonder if it is a good choice. If we change the configuration, it means that the integration test won’t cover the initial rebalance delay which is the default behavior. I think that we should rather keep it as it is.

**Comment by m1a2st** [comment]
Thanks for your comments, however I watch the kafka document 
> A new config, group.initial.rebalance.delay.ms was introduced. This config specifies the time, in milliseconds, that the GroupCoordinator will delay the initial consumer rebalance. The rebalance will be further delayed by the value of group.initial.rebalance.delay.ms as new members join the group, up to a maximum of max.poll.interval.ms. The default value for this is 3 seconds. During development and testing it might be desirable to set this to 0 in order to not delay test execution time.

According to the document, thus I consider that these tests are suitable for this config. WDYT?

**Comment by chia7712** [comment]
The test sends many `JoinGroupRequest` and each request needs to wait delay ("3" seconds). Also, the zk mode runs so fast due to initial_delay=0 (see following link)
https://github.com/apache/kafka/blob/e7317b37bf4a3ea5beaeb22a9b0707bdb16ebee7/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L352

Maybe we can reduce the delay from 3 seconds to 1.5 seconds? That not only keep the initial rebalance delay but also speed up the test.

**Comment by chia7712** [comment]
@m1a2st Could you please share the "before" and "after"?

**Comment by m1a2st** [comment]
- KafkaServerKRaftRegistrationTest 
  - Before:
CleanShot 2024-08-11 at 10 10 58@2x
  - After:
CleanShot 2024-08-11 at 11 05 23@2x
- ConsumerProtocolMigrationTest
  - Before:
CleanShot 2024-08-11 at 10 47 07@2x
  - After:
CleanShot 2024-08-11 at 11 08 06@2x
- JoinGroupRequestTest
  - Before:
CleanShot 2024-08-11 at 10 50 12@2x
  - After:
CleanShot 2024-08-11 at 11 08 54@2x
- ListGroupsRequestTest
  - Before:
CleanShot 2024-08-11 at 10 54 41@2x
  - After:
CleanShot 2024-08-11 at 11 09 39@2x
- OffsetFetchRequestTest
  - Before:
CleanShot 2024-08-11 at 10 56 28@2x
  - After:
CleanShot 2024-08-11 at 11 11 44@2x

- SyncGroupRequestTest
  - Before:
CleanShot 2024-08-11 at 10 58 48@2x
  - After:
CleanShot 2024-08-11 at 11 12 24@2x

**Comment by m1a2st** [comment]
According to the above test results, I consider we should modify `JoinGroupRequestTest`, `ListGroupsRequestTest` and `SyncGroupRequestTest` by add `group.initial.rebalance.delay.ms` config, and others should reduce the timeout from 120s to 30s

**Comment by chia7712** [comment]
> According to the above test results, I consider we should modify JoinGroupRequestTest, ListGroupsRequestTest and SyncGroupRequestTest by add group.initial.rebalance.delay.ms config

that makes sense to me

**Comment by chia7712** [comment]
@m1a2st could you please revise the title? thanks!

File: core/src/test/scala/integration/kafka/server/KafkaServerKRaftRegistrationTest.scala [hunk 2]
```diff
@@ -41,7 +41,7 @@ import scala.jdk.CollectionConverters._
  * failure paths is to use timeouts. See {@link unit.kafka.server.BrokerRegistrationRequestTest} for integration test
  * of just the broker registration path.
  */
-@Timeout(120)
+@Timeout(30)
```
Comment (chia7712) [inline core/src/test/scala/integration/kafka/server/KafkaServerKRaftRegistrationTest.scala]: please revert this unrelated change

File: core/src/test/scala/unit/kafka/server/ConsumerProtocolMigrationTest.scala [hunk 3]
```diff
@@ -28,7 +28,7 @@ import org.junit.jupiter.api.Assertions.assertEquals
 import org.junit.jupiter.api.Timeout
 import org.junit.jupiter.api.extension.ExtendWith
 
-@Timeout(120)
+@Timeout(30)
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/ConsumerProtocolMigrationTest.scala]: please revert this unrelated change

File: core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala [hunk 4]
```diff
@@ -29,7 +29,7 @@ import org.junit.jupiter.api.extension.ExtendWith
 
 import scala.jdk.CollectionConverters._
 
-@Timeout(120)
+@Timeout(30)
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/OffsetFetchRequestTest.scala]: please revert this unrelated change

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for updated PR

**Comment by m1a2st** [comment]
@chia7712, Thanks for your comments, revert it.

File: core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala [hunk 5]
```diff
@@ -51,7 +52,8 @@ class JoinGroupRequestTest(cluster: ClusterInstance) extends GroupCoordinatorBas
   @ClusterTest(serverProperties = Array(
     new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
     new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
-    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1"),
+    new ClusterConfigProperty(key = "group.initial.rebalance.delay.ms", value = "1"),
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala]: it should be "1000" rather than "1"

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this patch. please attach the results again as the updated config is incorrect

File: core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala (hunk 5 above)
Comment (m1a2st) [reply core/src/test/scala/unit/kafka/server/JoinGroupRequestTest.scala]: Yes, the unit is ms, I fix it.

**Comment by chia7712** [comment]
@m1a2st could you please share the results again?

**Comment by m1a2st** [comment]
The test results:
- `JoinGroupRequestTest`
CleanShot 2024-08-19 at 09 03 25@2x
- `ListGroupsRequestTest` 
CleanShot 2024-08-19 at 09 05 31@2x
- `SyncGroupRequestTest`
CleanShot 2024-08-19 at 09 08 34@2x

**Comment by m1a2st** [comment]
- `JoinGroupRequestTest`
    | test name | type | time(Before) | time(After) |
    | --- | --- | --- | --- |
    | JoinGroupRequestTest | | 3m 21s | 1m 43s |
    | testJoinGroupWithOldConsumerGroupProtocolAndNewGroupCoordinator | [1] Type=Raft-Isolated, MetadataVersion=4.0-IV1, Security=PLAINTEXT | 33s 328ms | 13s 557ms |
    | testJoinGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator | [1] Type=ZK, MetadataVersion=4.0-IV1, Security=PLAINTEXT | 2s 597ms | 13s 49ms |
    | testJoinGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator | [2] Type=Raft-Isolated, MetadataVersion=4.0-IV1, Security=PLAINTEXT | 32s 596ms | 12s 324ms |
    | testJoinGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator | [3] Type=Raft-Combined, MetadataVersion=4.0-IV1, Security=PLAINTEXT | 31s 986ms | 12s 329ms |

- `ListGroupsRequestTest`
    | test name | type | time(Before) | time(After) |
    | --- | --- | --- | --- |
    | ListGroupsRequestTest | | 7m 31s | 3m 18s |
    | testListGroupsWithNewConsumerGroupProtocolAndNewGroupCoordinator | [1] Type=Raft-Isolated, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 57s 215ms | 21s 428ms |
    | testListGroupsWithOldConsumerGroupProtocolAndNewGroupCoordinator | [1] Type=Raft-Isolated, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 55s 929ms | 19s 735ms |
    | testListGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator | [1] Type=ZK, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 1s 731ms | 20s 32ms |
    | testListGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator | [2] Type=Raft-Isolated, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 55s 362ms | 19s 34ms |
    | testListGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator | [3] Type=Raft-Combined, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 55s 3ms | 18s 945ms |

- `SyncGroupRequestTest`
    | test name | type | time(Before) | time(After) |
    | --- | --- | --- | --- |
    | SyncGroupRequestTest | | 2m 3s | 1m 4s |
    | testSyncGroupWithOldConsumerGroupProtocolAndNewGroupCoordinator | [1] Type=Raft-Isolated, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 20s 411ms | 8s 587ms |
    | testSyncGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator | [1] Type=ZK, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 1s 849ms | 8s 248ms |
    | testSyncGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator | [2] Type=Raft-Isolated, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 19s 753ms | 7s 728ms |
    | testSyncGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator | [3] Type=Raft-Combined, MetadataVersion=4.0-IVO, Security=PLAINTEXT | 19s 555ms | 7s 374ms |

**Comment by chia7712** [comment]
@m1a2st there are some tests getting slower. `testJoinGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator`, `testListGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator`, and `testSyncGroupWithOldConsumerGroupProtocolAndOldGroupCoordinator`. Do you know the root cause?

**Comment by m1a2st** [comment]
Becasue It is Zookeeper mode test, the config `group.initial.rebalance.delay.ms` in zookeeper mode default is 0 in https://github.com/apache/kafka/blob/4e1a015d36d42a595f7b74501e3b7a1510450a19/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L354 thus In these cases that we transfer to 1000 is slower than before. In KRaft mode it doesn't init the config value, thus the defulat value is 3000.

[setting]: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L354

**Comment by chia7712** [comment]
> thus In these cases that we transfer to 1000 is slower than before. In KRaft mode it doesn't init the config value, thus the defulat value is 3000.

makes sense. LGTM. will merge it later

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by chia7712** [comment]
@m1a2st could you please rebase code to trigger QA again? there are many failed tests. I feel they are unrelated, but double-check is not bad thing :)

**Comment by m1a2st** [comment]
I have rebased it :)

**Comment by chia7712** [comment]
@m1a2st please fix conflicts

**Comment by m1a2st** [comment]
Thanks for your comments. I resolve it.

**Comment by chia7712** [comment]
With merging this PR we can keep testing the global timeout for all IT (#16957)
